titus-server-master/src/main/java/com/netflix/titus/master/jobactivity/service/DefaultJobActivityPublisherService.java

```diff
@@ -0,0 +1,169 @@
+import com.netflix.titus.common.util.guice.annotation.Activator;
+import com.netflix.titus.common.util.guice.annotation.Deactivator;
+import com.netflix.titus.common.util.rx.ReactorExt;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+import reactor.core.scheduler.Schedulers;
+
+/**
+ * Implementation of a {@link JobActivityPublisherService} that consumes job and task events from the V3 engine
+ * and immediately propagates them to a store.
+ */
+@Singleton
+public class DefaultJobActivityPublisherService implements JobActivityPublisherService {
+    private static final Logger logger = LoggerFactory.getLogger(DefaultJobActivityPublisherService.class);
+
+    private static final String JOB_ACTIVITY_PUBLISHER_SCHEDULER = "JobActivityPublisherScheduler";
+
+    private final JobActivityPublisherStore publisher;
+    private final V3JobOperations v3JobOperations;
+    private Disposable jobUpdateEventDisposable;
+
+    private final TitusRuntime runtime;
+    private final JobActivityPublisherMetrics metrics;
+
+    private final Predicate<Job> jobActivityPublisherEnabledPredicate;
+
+    @Inject
+    public DefaultJobActivityPublisherService(JobActivityPublisherStore publisher,
+                                              V3JobOperations v3JobOperations,
+                                              @Named(FeatureRolloutPlans.JOB_ACTIVITY_PUBLISH_FEATURE) Predicate<JobDescriptor> jobActivityPublishPredicate,
+                                              TitusRuntime runtime) {
+        this.publisher = publisher;
+        this.v3JobOperations = v3JobOperations;
+        this.jobActivityPublisherEnabledPredicate = job -> jobActivityPublishPredicate.test(job.getJobDescriptor());
+        this.runtime = runtime;
+        this.metrics = new JobActivityPublisherMetrics(runtime.getRegistry());
+    }
+
+    @Activator
+    @Override
+    public void activate() {
+        logger.info("Starting job activity publisher");
+        jobUpdateEventDisposable = jobManagerStream().subscribe(
+                this::handleJobManagerEvent,
+                e -> {
+                    logger.error("Error in job activity publisher stream: ", e);
+                },
+                () -> {
+                    logger.error("Unexpected completion of job activity publisher stream");
+                }
+        );
+    }
+
+    @Deactivator
+    @Override
+    public void deactivate() {
+        jobUpdateEventDisposable.dispose();
+    }
+
+    @Override
+    public boolean isActive() {
+        return !jobUpdateEventDisposable.isDisposed();
+    }
+
+    private Flux<JobManagerEvent<?>> jobManagerStream() {
+        // The TitusRuntime emits stream metrics so we avoid explicitly managing them here
+        return ReactorExt.toFlux(runtime.persistentStream(v3JobOperations.observeJobs()))
+                .subscribeOn(Schedulers.newSingle(JOB_ACTIVITY_PUBLISHER_SCHEDULER));
```

Comment:
We have to set here back-pressure strategy.

Reply:
Done.